improver/snow_fraction.py

```diff
@@ -0,0 +1,165 @@
+                rates or accumulations of the same length and of compatible units.
+
+        Returns:
+            None
+
+        Raises:
+            ValueError:
+                If any of the criteria above are not met.
+
+        """
+        if len(input_cubes) != 2:
+            raise ValueError(
+                f"Expected exactly two input cubes, found {len(input_cubes)}"
+            )
+        rain_name, snow_name = self._get_input_cube_names(input_cubes)
+        self.rain = input_cubes.extract(rain_name).merge_cube()
+        self.snow = input_cubes.extract(snow_name).merge_cube()
+        self.snow.convert_units(self.rain.units)
+        if not spatial_coords_match(self.rain, self.snow):
+            raise ValueError("Rain and Snow cubes are not on the same grid")
+        if not self.rain.coord("time") == self.snow.coord("time"):
+            raise ValueError("Rain and Snow cubes do not have the same time coord")
+
+    @staticmethod
+    def _get_input_cube_names(input_cubes):
+        """
+        Identifies the rain and snow cubes from the presence of "rain" or "snow" in
+        the cube names.
+
+        Args:
+            input_cubes (iris.cube.CubeList):
+                The unsorted rain and snow cubes.
+
+        Returns:
+            tuple:
+                rain_name and snow_name, in that order.
+
+        """
+        cube_names = [cube.name() for cube in input_cubes]
+        try:
+            (rain_name,) = [x for x in cube_names if "rain" in x]
+            (snow_name,) = [x for x in cube_names if "snow" in x]
+        except ValueError:
+            raise ValueError(f"Could not find both rain and snow in {cube_names}")
+        if rain_name == snow_name:
+            raise ValueError(
+                f"Rain and Snow cubes must have different names, not {rain_name}"
+            )
+        return rain_name, snow_name
+
+    def _calculate_snow_fraction(self):
+        """
+        Calculates the snow fraction data and interpolates to fill in the missing points
+
+        Returns:
+            iris.cube.Cube:
+                Snow fraction cube
+
+        """
+        snow_fraction_cube = create_new_diagnostic_cube(
+            "snow_fraction",
+            "1",
+            template_cube=self.rain,
+            mandatory_attributes=generate_mandatory_attributes(
+                iris.cube.CubeList([self.rain, self.snow])
+            ),
+            data=self.snow.data / (self.rain.data + self.snow.data),
+        )
+
+        spatial_dims = [snow_fraction_cube.coord(axis=n).name() for n in ["y", "x"]]
+        snow_fraction_interpolated = iris.cube.CubeList()
+        for snow_fraction_slice in snow_fraction_cube.slices(spatial_dims):
+            snow_fraction_interpolated.append(
+                snow_fraction_slice.copy(
+                    interpolate_missing_data(snow_fraction_slice.data, method="nearest")
+                )
+            )
+        return snow_fraction_interpolated.merge_cube()
```

Comment:
I was concerned about nearest neighbour interpolation leading to snow diagnosis over the sea, or rain over the land when the temperatures of the two are quite different. However in your notebook you are not blending the snow fraction as a standalone diagnostic, instead it is multiplied by the precipitation rate and the rain/sleet/snow fields will go on to be blended. As such I think the problem goes away; you're filling in where the precip rate is zero and thus multiplying by zero when you use the rates.

This raises another question. If you are not blending or collapsing realizations on the snow fraction field before multiplying by the precip rate, why fill in the undefined regions at all? The interpolation step is a bit expensive and we multiply all the filled values by zero. In this regard it is equivalent to the [convective ratio calculation](https://github.com/metoppv/improver/blob/d33b4c4da5afe5402c2a3f19301969def86e408f/improver/convection.py#L403).

Reply:
We have to fill in the missing values in the snow-fraction field because we want to match a neighbourhood-smoothed precipitation probability field with it. I agree that nearest-neighbour infilling isn't quite as good as linear infilling, but the notebook shows that the differences are extremely slight. We also only care within a few pixels of a precipitating pixel. linear infilling is extremely expensive (cpu-time) so not worth using.

Reply:
Absolutely happy not to use linear. As you might be using neighbourhooded data it makes sense that you want a continuous field. As discussed, if this snow fraction approach is only to be used for global data on the global domain the above arguments will hold as long as we don't use neighbourhooding, but we may as well make it future proof with the infilling.